src/core/qgsmapsettings.cpp

```diff
@@ -423,6 +423,35 @@ QgsCoordinateTransform QgsMapSettings::layerTransform( const QgsMapLayer *layer
   return QgsCoordinateTransform( layer->crs(), mDestCRS, mTransformContext );
 }
 
+QgsRectangle QgsMapSettings::computeExtentForScale( const QgsPointXY &center, double scale ) const
+{
+  // Output width in inches
+  double outWIn = outputSize().width() / double( outputDpi() );
+
+  // Desired visible width (honouring scale)
+  double scaledWIn = outWIn * scale;
+
+  if ( mapUnits() == QgsUnitTypes::DistanceDegrees )
+  {
+    // Start with an 1x1 extent around the center
+    QgsRectangle ext( center.x() - 0.5, center.y() - 0.5, center.x() + 0.5, center.y() + 0.5 );
```

Comment:
I don't think always using 1 x 1 is appropriate here -- better to use a fraction of the existing width/height?